src/compiler/scala/tools/nsc/classpath/AggregateClassPath.scala

```diff
@@ -108,7 +107,7 @@ case class AggregateClassPath(aggregates: Seq[ClassPath]) extends ClassPath {
 
         if (existing.binary.isEmpty && entry.binary.isDefined)
           mergedEntries(index) = ClassAndSourceFilesEntry(entry.binary.get, existing.source.get)
-        if (existing.source.isEmpty && entry.source.isDefined)
+        else if (existing.source.isEmpty && entry.source.isDefined)
```

Comment:
What's the reasoning behind this change? I don't really understand the existing code to reason about the change... 😕

Reply:
I thought that they were mutually exclusive, but this was from the old branch so I have not looked in detail

I think existing has source or binary populated, and so does entry

Reply:
I believe that `entry.binary.isDefined && entry.source.isDefined` is always false, because `entry.isInstanceOf[ClassAndSourceFilesEntry]` is always fales. assuming we never wrap `AggregateClasspath`s.

The factory [`AggregateClassPath.apply`](https://github.com/scala/scala/blob/b9520211b22e53d4f80801b0f64cdb6275ac55cf/src/compiler/scala/tools/nsc/classpath/AggregateClassPath.scala#L137-L147) prevents that wrapping, but we don't actually guard against direct calls to `new AggregateClasspath`.

I'd suggest restoring the `else` for now to avoid the brain strain of figuring this out, as this implementation detail will likely change as we refactor towards a more performant design.

Reply:
removed